src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs

```diff
@@ -63,12 +63,136 @@ private BoundExpression BindIsPatternExpression(IsPatternExpressionSyntax node,
             DiagnosticBag diagnostics)
         {
             var type = (NamedTypeSymbol)this.BindType(node.Type, diagnostics);
+
+            var operators = type.GetMembers(WellKnownMemberNames.IsOperatorName);
+            if (operators.IsDefaultOrEmpty && ((CSharpParseOptions)node.SyntaxTree.Options).IsFeatureEnabled(MessageID.IDS_FeaturePatternMatching2))
+            {
+                // PROTOTYPE(patterns): As a temporary hack we recognize constructors and try to infer
+                // PROTOTYPE(patterns): a pattern-matching against properties based on constructor parameter names.
+                return BindInferredRecursivePattern(node, type, operand, operandType, hasErrors, diagnostics);
+            }
+            else
+            {
+                return BindUserDefinedRecursivePattern(node, type, operators, operand, operandType, hasErrors, diagnostics);
+            }
+        }
+
+        private BoundPattern BindUserDefinedRecursivePattern(
+            RecursivePatternSyntax node,
+            NamedTypeSymbol type,
+            ImmutableArray<Symbol> operators,
+            BoundExpression operand,
+            TypeSymbol operandType,
```

Comment:
We don't always have an operand (from source) in hand, for example in the case of a nested pattern.